scripts/pretty-printers/gdb/pretty_printers.py

```diff
@@ -236,14 +237,57 @@ def display_hint(self):
         return "string"
 
 
+def untypedef(type_obj):
+    if (type_obj.code == gdb.TYPE_CODE_REF or
+            type_obj.code == getattr(gdb, 'TYPE_CODE_RVALUE_REF', None)):
+
+        type_obj = type_obj.target()
+
+    if type_obj.code == gdb.TYPE_CODE_TYPEDEF:
+        type_obj = type_obj.strip_typedefs()
+
+    return type_obj
+
+
+def child_of_irept(val):
+    """ Use the irep pretty-printer if we're a child of irept. Based on the
+        assumption that all children will be using the sub/named_sub capabilities
+        of irept and have no other internal fields. """
+
+    type = untypedef(val.type)
+    if type is None or type.name is None:
+        return
+
+    if type.code == gdb.TYPE_CODE_STRUCT \
+        or type.code == gdb.TYPE_CODE_ENUM \
+        or type.code == gdb.TYPE_CODE_UNION:
+
+        hierarchy_types = set()
+        while type is not None:
+            if type.name is None or type.name in hierarchy_types:
+                break
+            else:
+                hierarchy_types.add(type.name)
+
+            for field in type.fields():
+                if field.is_base_class:
+                    type = untypedef(field.type)
+
+        if "irept" in hierarchy_types:
+            return IrepPrettyPrinter(val)
+
+    return None
+
+
 # If you change the name of this make sure to change install.py too.
 def load_cbmc_printers():
+    gdb.printing.register_pretty_printer(None, child_of_irept)
```

Comment:
:question: why none?

Reply:
So it's a 'global' pretty-printer, GDB allows you to attach pretty-printers to specific file objects, binaries etc, but I never got it working reliably. Making it global means any time something needs to be printed, it'll check to see if this matches.

Not precisely optimal, but right now I don't want to spend the time coaxing it to work (and the benefits are a big question-mark).

Reply:
Ohhh I see, for some reason I thought the second param was like the matcher.